lib/options.nix

```diff
@@ -333,5 +333,120 @@ rec {
+            '';
+
+            backend =
+              mkEnumFirstDefault
+                [
+                  "lz.n"
+                ]
+                ''
+                  The lazy-loading backend to use.
+                '';
+
+            # Spec loading:
+            enabled = mkStrLuaFnOr types.bool (lazyLoadDefaults.enabledInSpec or null) ''
+              When false, or if the function returns false, then ${originalName} will not be included in the spec.
+
+              Equivalence: lz.n => enabled; lazy.nvim => enabled
+            '';
+
+            priority = mkNullable types.number (lazyLoadDefaults.priority or null) ''
+              Only useful for start plugins (not lazy-loaded) to force loading certain plugins first.
+
+              Equivalence: lz.n => priority; lazy.nvim => priority
+            '';
+
+            # Spec setup
+            # Actions
+            beforeAll = mkLuaFn (lazyLoadDefaults.beforeAll or null) ''
+              Always executed before any plugins are loaded.
+
+              Equivalence: lz.n => beforeAll; lazy.nvim => init
+            '';
+
+            before = mkLuaFn (lazyLoadDefaults.before or null) ''
+              Executed before ${originalName} is loaded.
+
+              Equivalence: lz.n => before; lazy.nvim => None
+            '';
+
+            after = mkLuaFn (lazyLoadDefaults.after or null) ''
+              Executed after ${originalName} is loaded.
+
+              Equivalence: lz.n => after; lazy.nvim => config
+            '';
+
+            # Triggers
+            event = mkNullable triggerType (lazyLoadDefaults.event or null) ''
+              Lazy-load on event. Events can be specified as `BufEnter` or with a pattern like `BufEnter *.lua`
+
+              Equivalence: lz.n => event; lazy.nvim => event
+            '';
+
+            cmd = mkNullable triggerType (lazyLoadDefaults.cmd or null) ''
+              Lazy-load on command.
+
+              Equivalence: lz.n => cmd; lazy.nvim => cmd
+            '';
+
+            ft = mkNullable triggerType (lazyLoadDefaults.ft or null) ''
+              Lazy-load on filetype.
+
+              Equivalence: lz.n => ft; lazy.nvim => ft
+            '';
+
+            keys = mkListOf (types.attrsOf types.anything) (lazyLoadDefaults.keys or null) ''
+              Lazy-load on key mapping.
+
+              Equivalence: lz.n => keys; lazy.nvim => keys
+            '';
+
+            colorscheme = mkNullable triggerType (lazyLoadDefaults.colorscheme or null) ''
+              Lazy-load on colorscheme.
+
+              Equivalence: lz.n => colorscheme; lazy.nvim => None
+            '';
+
+            extraSettings = mkSettingsOption {
+              description = ''
+                Extra settings to pass to the lazy loader backend.
+              '';
+              example = {
+                dependencies = {
+                  __unkeyed-1 = "nvim-lua/plenary.nvim";
+                  lazy = true;
+                };
+              };
```

Comment:
It would maybe be better to only support a `backendSettings` option of type `attrsOf anything`, that will be passed to the code that implements the lazy loading provider.
I'm not confident that we can find a common API for all possible lazy loading schemes

Reply:
That makes sense. I was just changing the options nesting path and this would fit nicely. We can wrap all the `backendSettings` in a `mkSettingsOption` with the known options specified for documentation.

Reply:
> I'm not confident that we can find a common API for all possible lazy loading schemes

Perhaps. But I think we will need _some_ level of abstraction even if it is just knowing where to put our setup call.

That could look like a `setupOptionLoc` option with a default based on which backend is enabled?

E.g.
```nix
setupOptionLoc = mkOption {
  type = listOf str; # maybe a non-merging list instead?
  default =
    if config.pligins.lz-n.enable then
      [ "after" ]
    else if config.lazy.enable then
      [ "config" ]
    else
      throw "setupOptionLoc cannot be used when no lazyloader is enabled";
  defaultText = literalMD "TODO";
}

Reply:
I was thinking of a function that takes as an input all the required information.... But that will end up in an infinite recursion.

We *do* need an enumeration of supported backends, but we may not need more than that